### PR TITLE
feat(analytics): adding metric api for dispute analytics

### PR DESCRIPTION
**Pull request by harsh-sharma-juspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
adding metric api for dispute analytics
Http Method: ```POST```
URL Path ```/analytics/v1/metric/dispute```

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Hit the URL given below and should get sample response mentioned in the comment

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by harsh-sharma-juspay** [comment]
Curl:
```
curl --location 'http://localhost:8080/analytics/v1/metrics/disputes' \
--header 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:121.0) Gecko/20100101 Firefox/121.0' \
--header 'Accept: */*' \
--header 'Accept-Language: en-US,en;q=0.5' \
--header 'Accept-Encoding: gzip, deflate, br' \
--header 'Referer: https://app.hyperswitch.io/' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiM2NhODdjNTUtYTI1YS00YjRmLWE3MWUtMjM4NzU1MTA5NDBlIiwibWVyY2hhbnRfaWQiOiJtZXJjaGFudF8xNzA4MDc1MjcyIiwicm9sZV9pZCI6Im9yZ19hZG1pbiIsImV4cCI6MTcwOTA1NDk0OSwib3JnX2lkIjoib3JnX0s2TkNFa0R1dHNqNUFuazJOVjl6In0.cYJjEMJG4-hs74SdFeF4_wAKBbikvlpW3f-8aP_yioQ' \
--header 'api-key: hyperswitch' \
--header 'Connection: keep-alive' \
--data '[
    {
        "timeRange": {
            "startTime": "2024-02-17T18:30:00Z",
            "endTime": "2024-02-25T17:34:00Z"
        },
        "groupByNames": [
            "connector"
        ],
        "source": "BATCH",
        "metrics": [
            "dispute_status_metric"
        ],
        "delta": true
    }
]'
```

**Comment by harsh-sharma-juspay** [comment]
```
{
    "queryData": [
        {
            "disputes_challenged": 0,
            "disputes_won": 0,
            "disputes_lost": 0,
            "total_amount_disputed": null,
            "total_dispute_lost_amount": null,
            "total_dispute": 2,
            "dispute_stage": null,
            "connector": "checkout",
            "dispute_status": "dispute_opened",
            "time_range": {
                "start_time": "2024-02-17T18:30:00.000Z",
                "end_time": "2024-02-25T17:34:00.000Z"
            },
            "time_bucket": "2024-02-17 18:30:00"
        }
    ],
    "metaData": [
        {
            "current_time_range": {
                "start_time": "2024-02-17T18:30:00.000Z",
                "end_time": "2024-02-25T17:34:00.000Z"
            }
        }
    ]
}
```

**Comment by harsh-sharma-juspay** [comment]
Screenshot 2024-02-26 at 11 42 50 AM

**Review by ivor-juspay — DISMISSED** [review]

File: crates/analytics/src/disputes/accumulators.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        if self.total <= 0 {
+            Some((None, None, None, None))
+        } else {
+            Some((
+                u64::try_from(self.challenged_count).ok(),
+                u64::try_from(self.won_count).ok(),
+                u64::try_from(self.lost_count).ok(),
+                u64::try_from(self.total).ok(),
+            ))
+        }
```
Comment (lsampras) [inline crates/analytics/src/disputes/accumulators.rs:83]: is this rate or count?
should it be `challenged_count/total` instead

File: crates/api_models/src/analytics/disputes.rs [hunk 2]
```diff
@@ -64,3 +69,75 @@ impl From<DisputeMetrics> for NameDescription {
         }
     }
 }
+
+use crate::enums::DisputeStage;
```
Comment (lsampras) [inline crates/api_models/src/analytics/disputes.rs:73]: move this import to start of file,
would prefer having these imports in the disputes file maybe..

File: crates/analytics/src/disputes/accumulators.rs (hunk 1 above)
Comment (harsh-sharma-juspay) [reply crates/analytics/src/disputes/accumulators.rs:83]: it have total count as well so if rate have to be display FE can convert count to rate percentage

File: crates/analytics/src/disputes/metrics/dispute_status_metric.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        filters: &DisputeFilters,
+        granularity: &Option<Granularity>,
+        time_range: &TimeRange,
+        pool: &T,
+    ) -> MetricsResult<Vec<(DisputeMetricsBucketIdentifier, DisputeMetricRow)>>
+    where
+        T: AnalyticsDataSource + super::DisputeMetricAnalytics,
+    {
+        let mut query_builder = QueryBuilder::new(AnalyticsCollection::Dispute);
+        // let dimensions = dimensions.to_vec();
```
Comment (lsampras) [inline crates/analytics/src/disputes/metrics/dispute_status_metric.rs:40]:
```suggestion
```

File: crates/analytics/src/disputes/metrics/total_amount_disputed.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        granularity: &Option<Granularity>,
+        time_range: &TimeRange,
+        pool: &T,
+    ) -> MetricsResult<Vec<(DisputeMetricsBucketIdentifier, DisputeMetricRow)>>
+    where
+        T: AnalyticsDataSource + super::DisputeMetricAnalytics,
+    {
+        let mut query_builder: QueryBuilder<T> = QueryBuilder::new(AnalyticsCollection::Dispute);
+
+        for dim in dimensions.iter() {
```
Comment (lsampras) [inline crates/analytics/src/disputes/metrics/total_amount_disputed.rs:41]:
```suggestion
        for dim in dimensions {
```

**Review by ivor-juspay — APPROVED** [review]

**Review by lsampras — APPROVED** [review]
rest looks good

**Review by lsampras — APPROVED** [review]

**Review by ivor-juspay — APPROVED** [review]

**Pull request merged**